### PR TITLE
add log msg when event recvr firehose buffer is full

### PR DESCRIPTION
**Pull request by himanshug**
indicates if/when realtime tasks are not being able to keep up with ingestion load and requests are stuck or busy waiting here.

### CODE REVIEW

**Comment by gianm** [comment]
This will log two messages a second if the buffer is full… seems like a lot, perhaps add a throttle (up to you)?

👍  otherwise though

**Comment by himanshug** [comment]
@gianm i guess it will be fine... printing of that log is really an exceptional situation... if that log is getting printed too much then that is an indication that something bad will eventually happen and more partitions,  tunings, hardware etc etc is needed to support the input data rate.

**Comment by gianm** [comment]
I just realized that since this is called by jetty threads, it will potentially log a lot more often than twice a second. Also this situation is not really that bad if it's a temporary situation (perhaps buffer is full due to a data rate spike/backlog, and you have QoS enabled to protect queries so they will be OK).

So for those two reasons I do think we should have a throttle here.

**Comment by himanshug** [comment]
i did not factor in the multiple threads too, let me throttle it to print every 10 retries instead of each one

**Comment by gianm** [comment]
IMO a time based throttle is best, like at most one message every 10 seconds or something.

**Comment by himanshug** [comment]
10/20 retries would equal to 5/10 secs as wait time is hardcoded to be 500ms..

or, do you mean once every 10 secs across all threads, that would need some global variables to be kept, not sure if we want to go that route ?

**Comment by himanshug** [comment]
or instead of "log" we can introduce a new metric for this case

e.g. ingest/buffer/failCount  and update it on each failed "add" call?

**Comment by gianm** [comment]
I meant once every 10 seconds across all threads, using an AtomicLong in the firehose. A metric works too. Metrics are also shared AtomicLongs so I think they're kinda similar in that regard.

**Comment by himanshug** [comment]
@gianm updated

File: server/src/main/java/io/druid/segment/realtime/firehose/EventReceiverFirehoseFactory.java [hunk 1]
```diff
@@ -298,6 +298,13 @@ public void addRows(Iterable<InputRow> rows) throws InterruptedException
         boolean added = false;
         while (!closed && !added) {
           added = buffer.offer(row, 500, TimeUnit.MILLISECONDS);
+          if (!added) {
+            long currTime = System.currentTimeMillis();
```
Comment (gianm) [inline server/src/main/java/io/druid/segment/realtime/firehose/EventReceiverFirehoseFactory.java]: minor race here, would be better as:

```java
long lastTime = lastBufferAddFailMsgTime.get();
if (currTime - lastTime > 10000 && lastBufferAddFailMsgTime.compareAndSet(lastTime, currTime)) {
  log.warn("Failed to add event to buffer with current size [%s] . Retrying...", buffer.size());
}
```

Comment (himanshug) [inline server/src/main/java/io/druid/segment/realtime/firehose/EventReceiverFirehoseFactory.java]: done

**Comment by fjy** [comment]
👍

**Comment by pjain1** [comment]
👍

**Pull request merged**